File: Linked_List_Cycle/hasCycle.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
        
def has_cycle(head: ListNode) -> ListNode | None:
    slow = fast = head
    
    while fast and fast.next:
        slow = slow.next
        fast = fast.next.next
        if slow is fast:
            return slow
    return None

def find_cycle_entry(head: ListNode) -> ListNode | None:
    meet = has_cycle(head)
    if not meet:
        return None
    
    p1, p2 = head, meet
    while p1 is not p2:
        p1 = p1.next
        p2 = p2.next
    return p1

def break_cycle(entry: ListNode) -> None:
    cur = entry
    while cur.next is not entry:
        cur = cur.next
    cur.next = None

if __name__ == "__main__":
    head = ListNode(1)
    n2 = ListNode(2)
    n3 = ListNode(3)
    
    head.next = n2
    n2.next = n3
    n3.next = head
    
    print("Has cycle:", bool(has_cycle(head)))
    
    entry = find_cycle_entry(head)
    if entry:
        break_cycle(entry)
        
    print("Cycle broken:", bool(has_cycle(head)))

File: Linked_List_Cycle/hasCycle.rs
use std::rc::Rc;
use std::cell::RefCell;

type Link = Option<Rc<RefCell<ListNode>>>;

#[derive(Debug)]
struct ListNode {
    val: i32,
    next: Link,
}

fn has_cycle(head: &Link) -> Link {
    let mut slow = head.clone();
    let mut fast = head.clone();
    
    while fast.is_some() && fast.as_ref().unwrap().borrow().next.is_some() {
        slow = slow.unwrap().borrow().next.clone();
        fast = fast
            .unwrap()
            .borrow()
            .next
            .as_ref()
            .unwrap()
            .borrow()
            .next
            .clone();
        
        if let (Some(s), Some(f)) = (&slow, &fast) {
            if Rc::ptr_eq(s, f) {
                return slow;
            }
        }
    }
    None
}

fn find_cycle_entry(head: &Link) -> Link {
    let meet = has_cycle(head)?;
    let mut p1 = head.clone();
    let mut p2 = Some(meet);
    
    while let (Some(a), Some(b)) = (p1.clone(), p2.clone()) {
        if Rc::ptr_eq(&a, &b) {
            return p1;
        }
        p1 = a.borrow().next.clone();
        p2 = b.borrow().next.clone();
    }
    None
}

fn break_cycle(entry: &Rc<RefCell<ListNode>>) {
    let mut cur = entry.clone();
    
    loop {
        let next = cur.borrow().next.clone().unwrap();
        if Rc::ptr_eq(&next, entry) {
            cur.borrow_mut().next = None;
            break;
        }
        cur = next;
    }
}

fn main() {
    let n1 = Rc::new(RefCell::new(ListNode { val: 1, next: None }));
    let n2 = Rc::new(RefCell::new(ListNode { val: 2, next: None }));
    let n3 = Rc::new(RefCell::new(ListNode { val: 3, next: None }));
    
    let head = Some(n1.clone());
    
    println!("Has cycle: {}", has_cycle(&head).is_some());
    
    if let Some(entry) = find_cycle_entry(&head) {
        break_cycle(&entry);
    }
    
    println!("Cycle broken: {}", has_cycle(&head).is_some());
}
